# Part 1 - Data Cleaning

In [1]:
#importing required packages for this notebook
import pandas as pd

In [2]:
df = pd.read_csv('data/fake_job_postings.csv')
df.shape

(17880, 18)

In [3]:
df.head(3)

job_id                                      title          location  \
0       1                           Marketing Intern  US, NY, New York   
1       2  Customer Service - Cloud Video Production    NZ, , Auckland   
2       3    Commissioning Machinery Assistant (CMA)     US, IA, Wever   

  department salary_range                                    company_profile  \
0  Marketing          NaN  We're Food52, and we've created a groundbreaki...   
1    Success          NaN  90 Seconds, the worlds Cloud Video Production ...   
2        NaN          NaN  Valor Services provides Workforce Solutions th...   

                                         description  \
0  Food52, a fast-growing, James Beard Award-winn...   
1  Organised - Focused - Vibrant - Awesome!Do you...   
2  Our client, located in Houston, is actively se...   

                                        requirements  \
0  Experience with content management systems a m...   
1  What we expect from you:Your key responsibilit...   
2  Implement pre-commissioning and commissioning ...   

                                            benefits  telecommuting  \
0                                                NaN              0   
1  What you will get from usThrough being part of...              0   
2                                                NaN              0   

   has_company_logo  has_questions employment_type required_experience  \
0                 1              0           Other          Internship   
1                 1              0       Full-time      Not Applicable   
2                 1              0             NaN                 NaN   

  required_education                   industry          function  fraudulent  
0                NaN                        NaN         Marketing           0  
1                NaN  Marketing and Advertising  Customer Service           0  
2                NaN                        NaN               NaN           0

## Data Dicitonary
There are 17880 rows with 18 features.

|Column/Feature|Discription|
|---|---|
|job_id|Unique Job ID|
|title|The title of the job ad entry.|
|location|Geographical location of the job ad.|
|department|Corporate department (e.g. sales).|
|salary_range|Indicative salary range (e.g. $50,000-$60,000)|
|company_profile|A brief company description.|
|description|The details description of the job ad.|
|requirment|Enlisted requirements for the job opening.|
|benefits|Enlisted offered benefits by the employer.|
|telecommuting|True for telecommuting positions.|
|has_company_logo|True if company logo is present.|
|has_questions|True if screening questions are present.|
|employment_type|Full-type, Part-time, Contract, etc.|
|required_experience|Executive, Entry level, Intern, etc.|
|required_education|Doctorate, Master’s Degree, Bachelor, etc.|
|industry|Automotive, IT, Health care, Real estate, etc.|
|function|Consulting, Engineering, Research, Sales etc.|
|fradulent|target - Classification attribute.|

**Target Variable** = fradulent (1 or 0) with 1 being fradulent

In [4]:
#Checking our Data Types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17880 entries, 0 to 17879
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   job_id               17880 non-null  int64 
 1   title                17880 non-null  object
 2   location             17534 non-null  object
 3   department           6333 non-null   object
 4   salary_range         2868 non-null   object
 5   company_profile      14572 non-null  object
 6   description          17879 non-null  object
 7   requirements         15185 non-null  object
 8   benefits             10670 non-null  object
 9   telecommuting        17880 non-null  int64 
 10  has_company_logo     17880 non-null  int64 
 11  has_questions        17880 non-null  int64 
 12  employment_type      14409 non-null  object
 13  required_experience  10830 non-null  object
 14  required_education   9775 non-null   object
 15  industry             12977 non-null  object
 16  func

In [5]:
#Check percentage of data missing for each feature/column
df.isna().sum()/len(df)

job_id                 0.000000
title                  0.000000
location               0.019351
department             0.645805
salary_range           0.839597
company_profile        0.185011
description            0.000056
requirements           0.150727
benefits               0.403244
telecommuting          0.000000
has_company_logo       0.000000
has_questions          0.000000
employment_type        0.194128
required_experience    0.394295
required_education     0.453300
industry               0.274217
function               0.361018
fraudulent             0.000000
dtype: float64

In [6]:
#Checking for unique elements for each column
df.nunique()

job_id                 17880
title                  11231
location                3105
department              1337
salary_range             874
company_profile         1709
description            14801
requirements           11968
benefits                6205
telecommuting              2
has_company_logo           2
has_questions              2
employment_type            5
required_experience        7
required_education        13
industry                 131
function                  37
fraudulent                 2
dtype: int64

Just from a quick glance my data, it seems that there is quite a few features that have a lot of missing rows. As such, **I will delete the various columns:**

1. job_id because my DataFrame already has a built in index. 
2. salary_range because around 84% of the data is missing
3. department because around 65% of the data is missing
4. benefits because 40% of the data is missing
5. company_profile because I want to combine the description + requirements columns to one features, in order to perform a tfidf vectorizer on it later on. 

The rest of the columns will be filled out in a methedolical order. 

In [7]:
# Always good practice to make a copy of your dataframe ever so often,
# so you can roll back your mistakes much easier than rerunning your whole kernal again.
df_2 = df.copy()
df_2 = df_2.drop(labels = ['job_id','salary_range',
                    'department','benefits',
                    'company_profile'], axis = 1) #axis = 1 to refer droping columns

In [8]:
df_2.tail(3)

title           location  \
17877  Project Cost Control Staff Engineer - Cost Con...    US, TX, Houston   
17878                                   Graphic Designer      NG, LA, Lagos   
17879                         Web Application Developers  NZ, N, Wellington   

                                             description  \
17877  Experienced Project Cost Control Staff Enginee...   
17878  Nemsia Studios is looking for an experienced v...   
17879  Who are we?Vend is an award winning web based ...   

                                            requirements  telecommuting  \
17877  At least 12 years professional experience.Abil...              0   
17878  1. Must be fluent in the latest versions of Co...              0   
17879  We want to hear from you if:You have an in-dep...              0   

       has_company_logo  has_questions employment_type required_experience  \
17877                 0              0       Full-time                 NaN   
17878                 0              1        Contract      Not Applicable   
17879                 1              1       Full-time    Mid-Senior level   

      required_education           industry     function  fraudulent  
17877                NaN                NaN          NaN           0  
17878       Professional     Graphic Design       Design           0  
17879                NaN  Computer Software  Engineering           0

Filling missing values for **employment_type**, **required_experience**, **required_education**, **industry**, **function** using the pandas bfill function. I did this because these features had the fewest unique elements for a non-binary feature.

> *Pandas bfill is a function that is used with the fillna function to back fill the values in a dataframe. Thus, if there is a NaN cell then bfill will replace that NaN value with the next row or column based on the axis equaling to 0 or 1.*

In [9]:
df_2['employment_type'] = df_2['employment_type'].bfill(axis=0)
df_2['required_experience'] = df_2['required_experience'].bfill(axis = 0)
df_2['required_education'] = df_2['required_education'].bfill(axis = 0)
df_2['industry'] = df_2['industry'].bfill(axis = 0)
df_2['function'] = df_2['function'].bfill(axis = 0)

Next step is to append the *description* column and *requirments* column together into one column. However, before I do this, **I want to avoid the NaN values in both of these columns.** In order to do so since there is a small number of missing rows in the description column, I will drop those rows first. From there, I will fill in all NaN values in the *requirments* column with " " aka. blank string. 

In addition I will drop duplicated description columns as well, prior to the great concatenation. 

In [10]:
# Make Dataframe copy
df_3 = df_2.copy()

# Keeping non NaN rows in my new dataframe
df_3 = df_3[df_3['description'].notna()]

# Replacing NaNs with an empty string.
#df_3 = df_3.replace(np.nan, '', regex = True)

In [11]:
# For good measure let's drop any other Nans 
df_3 = df_3.dropna(axis = 0, how = 'any')

In [12]:
print(f'We currenlty have {len(df_3)} rows. However, let\'s drop duplicates and compare.')

We currenlty have 15014 rows. However, let's drop duplicates and compare.


In [13]:
# drop duplicates
df_3 = df_3.drop_duplicates(keep = 'first')

In [14]:
df_3.isna().sum()/len(df)

title                  0.0
location               0.0
description            0.0
requirements           0.0
telecommuting          0.0
has_company_logo       0.0
has_questions          0.0
employment_type        0.0
required_experience    0.0
required_education     0.0
industry               0.0
function               0.0
fraudulent             0.0
dtype: float64

In [15]:
print(f'After dropping duplicates we have {len(df_3)} rows left. It seems there were 178 duplicate rows.')

After dropping duplicates we have 14836 rows left. It seems there were 178 duplicate rows.


In [16]:
# Make copy
df_4 = df_3.copy()

#concatenating our description and requirments columns
df_4['description'] = df_4['description'] + ' ' + df_4['requirements']
del df_4['requirements']

In [17]:
#Clean DataFrame
df_clean = df_4.copy()

display(df_clean.head(7))
print(df_clean.shape)

title               location  \
0                           Marketing Intern       US, NY, New York   
1  Customer Service - Cloud Video Production         NZ, , Auckland   
2    Commissioning Machinery Assistant (CMA)          US, IA, Wever   
3          Account Executive - Washington DC     US, DC, Washington   
4                        Bill Review Manager     US, FL, Fort Worth   
6                      Head of Content (m/f)         DE, BE, Berlin   
7           Lead Guest Service Specialist     US, CA, San Francisco   

                                         description  telecommuting  \
0  Food52, a fast-growing, James Beard Award-winn...              0   
1  Organised - Focused - Vibrant - Awesome!Do you...              0   
2  Our client, located in Houston, is actively se...              0   
3  THE COMPANY: ESRI – Environmental Systems Rese...              0   
4  JOB TITLE: Itemization Review ManagerLOCATION:...              0   
6  Your Responsibilities: Manage the English-spea...              0   
7  Who is Airenvy?Hey there! We are seasoned entr...              0   

   has_company_logo  has_questions employment_type required_experience  \
0                 1              0           Other          Internship   
1                 1              0       Full-time      Not Applicable   
2                 1              0       Full-time    Mid-Senior level   
3                 1              0       Full-time    Mid-Senior level   
4                 1              1       Full-time    Mid-Senior level   
6                 1              1       Full-time    Mid-Senior level   
7                 1              1       Full-time           Associate   

          required_education                             industry  \
0          Bachelor's Degree            Marketing and Advertising   
1          Bachelor's Degree            Marketing and Advertising   
2          Bachelor's Degree                    Computer Software   
3          Bachelor's Degree                    Computer Software   
4          Bachelor's Degree               Hospital & Health Care   
6            Master's Degree                         Online Media   
7  High School or equivalent  Information Technology and Services   

               function  fraudulent  
0             Marketing           0  
1      Customer Service           0  
2                 Sales           0  
3                 Sales           0  
4  Health Care Provider           0  
6            Management           0  
7      Customer Service           0

(14836, 12)


After cleaning our data, we went from 17,880 rows with 18 features to 14,836 rows with 12 features.

In [18]:
df_clean.to_csv('data/cleaned_data.csv')